asim: port first batch of testing framework changes

tbg opened the pull request:
This ports a few improvements from #149531 to master:

- ability to combine multiple load generators
- improved range creation (placement_type option, min/max key, ...)
- ability to disable split/lease/replicate queue
- plots show initial and final values

I also cleaned up the args handling to make sure that we don't silently skip args not supported by the test harness. This was helpful to find what still needed porting.

A first datadriven test file was moved over; the CPU/write tracking is still unsupported, so I commented it out with TODOs to re-enable once possible. Unfortunately, introducing this will be a bigger yak shave, since this data is communicated through the `roachpb` protos, and we may not want to change them without also first adding the code that populates these fields correctly in production builds.

Epic: CRDB-25222

wenyihu6 committed:
asim: simplify logic in history.ShowRecordedValueAt

Previously, we introduced support for showing initial and final metric values in
3da645a0aff3d46ea5b0be9e79d5b1c4837deedf, including an additional sorting step.
However, the store metrics were already provided in sorted order. This commit
removes the unnecessary sorting to simplify the logic.

Epic: none
Release note: none

tbg committed:
gen: add MultiLoad

We use the syntax `replace=<bool>` instead of `add_to_existing` since it
appears that the default intention is to add to the existing workload.

tbg committed:
asim: add queue settings

Allow disabling the various queues.

tbg committed:
asim: add replica_placement, min_key, max_key

and make other changes that were required to port the
`load_distribution_movement_disabled` datadriven file from #149531. Not
everything in that file works yet; TODOs have been added.

Co-authored-by: wenyihu6 <wenyi@cockroachlabs.com>

tbg committed:
tests: add and use scanMustExist

scanIfExists and scanMustExist consume the args the match on. An assertion was
added to make sure we don't leave any args unconsumed, which would constitute a
bug in testing (and a common kind in datadriven tests).

tbg committed:
tests: fix set_liveness doc

tbg committed:
history: also print sum, not just stddev

tbg committed:
asim: remove extra newline

This significantly tightens up the plot graphs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149582)
<!-- Reviewable:end -->

wenyihu6 committed:
asim: remove unused keyspace from BaseRanges and rand framework

Previously, we add min/max key to BaseRanges in
3da645a0aff3d46ea5b0be9e79d5b1c4837deedf,
which made keyspace field unused. This commit removes the unused
keyspace field from BaseRanges. Additionally, it introduces
min/max keys for the rand framework. Note that some rand tests now
fail due to a simulator assumption that the first range must start
with the key 0. A future commit will explain and address it.

Epic: none
Release note: none

wenyihu6 committed:
asim: fix under-replication in rand testing

Previously, the simulator assumed that the LoadRangeInfo will include a range
with min key as 0 (the first range created by initFirstRange relies on
LoadRangeInfo to handle up-replication). When rand tests overrode this
assumption, it led to under replication and failed assertion. This commit goes
around this by setting defaultMinKey = 0 and disallowing the rand framework from
overriding. We should fix the simulator assumption eventually.

Epic: none
Release note: none

wenyihu6 committed:
asim: remove unused keyspace for rand testing

Previously, we added min/max key with staticOptionSettings, making keyspace
unused. This commit finishes the cleanup by removing the unused keyspace field.

Epic: none
Release note: none

wenyihu6 committed:
asim: move NewStateWith* helpers into separate file

Previously, NewStateWith* helpers were with simulator logic, even though
they were only used for unit tests (not datadriven tests). This commit
moves them into a separate file to make it clearer that they are purely testing
utilities and to clear confusion when changing code.

Epic: none
Release note: none

wenyihu6 committed:
asim: sue DefaultSpanConfigWithRF in more places

Previously, we added DefaultSpanConfigWithRF as a helper function. This commit
makes use of it at more places.

Epic: none
Release note: none

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Preemptively approve to unblock merging and porting. I addressed my comments in my commits. Feel free to check if they look good to you and fix up as you see fit. 

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-:-OUbeA2y4osOCHR-rN7I:b-g7lvf5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149582)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/rand_test.go` line 186 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-OUcoZMyCRLGrOMocrid:-OUcoZMz1K4aB4xP2tRH:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b3f1437f9aef31de5f6354c83ed875efeec5c3c6/pkg/kv/kvserver/asim/tests/rand_test.go#L186)):*
> ```Go
> 				scanIfExists(t, d, "max_block", &staticOptionSettings.maxBlock)
> 				if hasMinKey := scanIfExists(t, d, "min_key", &staticOptionSettings.minKey); hasMinKey && staticOptionSettings.minKey != int64(0) {
> 					// TODO(wenyihu6): the simulator assumes the first range starts at 0.
> ```

:thumbsup: 
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_splitting` line 66 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-OUcpHhu1GuZQZOLSYlS:-OUcpHhu1GuZQZOLSYlT:b-uluoo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c776f50c2fbe7944efddbdde9dfa48853cba87a5/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_splitting#L66)):*
> ```
> plot stat=replicas sample=4
> ----
>  15.00 ┤                                ╭──────────────────────────────────────────────
> ```

Isn't it surprising that this graph changed? That would indicate to me that something wasn't sorted (or that we at least made some rearrangement that now is no longer carried out). Do you know where this comes from?
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter` line 0 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-OUcpbeK0D4r5skSZUub:-OUcpbeK0D4r5skSZUuc:b-tyddga) ([raw file](https://github.com/cockroachdb/cockroach/blob/c776f50c2fbe7944efddbdde9dfa48853cba87a5/pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter#L0)):*
I wonder why Reviewable considers this file binary. I've also had similar issues in Goland, some of the testdata files are more or less randomly considered binary and I have to force Goland to open them as text.


<!-- Sent from Reviewable.io -->

tbg commented:
Looks like some engflow fluke?
<img width="1340" alt="image" src="https://github.com/user-attachments/assets/03dd34ac-9059-4d43-bdfd-60064de74b22" />

The job is still pending somehow, I'll give it some time and try again later.

tbg commented:
bors r+

Found the issue: a way that I added to avoid an unused lint caused an NPE at init time.
Still weird that this failed so opaquely.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149582)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_splitting` line 66 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-OUcpHhu1GuZQZOLSYlS:-OUlLras8GCF2yqhUuC3:b-aefaex) ([raw file](https://github.com/cockroachdb/cockroach/blob/c776f50c2fbe7944efddbdde9dfa48853cba87a5/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_splitting#L66)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't it surprising that this graph changed? That would indicate to me that something wasn't sorted (or that we at least made some rearrangement that now is no longer carried out). Do you know where this comes from?
</blockquote></details>

This happened because we introduced `replace` in the process but forgot to set `replace = true` here. The default behavior is replace = false which adds appends a new load generator. I put up a PR to set it correctly and confirmed that the output now matches what you expected. https://github.com/cockroachdb/cockroach/pull/149847/files.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter` line  at r13](https://reviewable.io/reviews/cockroachdb/cockroach/149582#-OUcpbeK0D4r5skSZUub:-OUlMiluDdi4kmAC48jv:b-4gaulp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c776f50c2fbe7944efddbdde9dfa48853cba87a5/pkg/kv/kvserver/asim/tests/testdata/non_rand/one_voter#L)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I wonder why Reviewable considers this file binary. I've also had similar issues in Goland, some of the testdata files are more or less randomly considered binary and I have to force Goland to open them as text.
</blockquote></details>

I’ve run into that too and found it annoying. I can put up a PR to rename the test data files to `.txt`, and do the same in the prototype branch if we want.


<!-- Sent from Reviewable.io -->
